### PR TITLE
Search content in emails

### PR DESCRIPTION
**Pull request by seiferla**

### CODE REVIEW

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/content/GenericSearchEmailService.java [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package de.dhbw.karlsruhe.students.mailflow.core.application.email.searchemail.content;
+
+public class GenericSearchEmailService implements SearchEmailUseCase{}
```
Comment (jens1o) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/content/GenericSearchEmailService.java:3]:
```suggestion
public class GenericSearchEmailService implements SearchEmailUseCase {}
```

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/content/SearchContentEmailUseCase.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.Email;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.exceptions.MailboxLoadingException;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.exceptions.MailboxSavingException;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.value_objects.Address;
+import java.util.List;
+
+public interface SearchContentEmailUseCase {
+
+  List<Email> searchContentInEmails(String content, Address address)throws MailboxSavingException, MailboxLoadingException;
```
Comment (jens1o) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/content/SearchContentEmailUseCase.java:11]:
```suggestion
  List<Email> searchContentInEmails(String content, Address address) throws MailboxSavingException, MailboxLoadingException;
```

**Review by jens1o — COMMENTED** [review]
just some reformatting changes for now, I see that some functionality isn't fully implemented, yet.

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/GenericSearchEmailService.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  public List<Email> searchEmails(Address address)
+      throws MailboxSavingException, MailboxLoadingException {
+    Mailbox mailbox;
+    List<Email> emailList = new ArrayList<>();
+    for (MailboxType type : MailboxType.values()) {
+      mailbox = mailboxRepository.findByAddressAndType(address, type);
+      emailList.addAll(mailbox.getEmailList());
+    }
+    return emailList;
+  }
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/GenericSearchEmailService.java:32]: This functionality should be in the Services which extend the GenericSearchEmailService. I would also suggest renaming it to AbstractSearchEmailSerivce. 

This abstract class could provide functionality like `public boolean containsCaseInsensitive(String emailString, String searchTerm)`. The extending Services could use this functionality while implementing searchEmails on its own

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/GenericSearchEmailService.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class GenericSearchEmailService implements SearchEmailUseCase {
+
+  private final MailboxRepository mailboxRepository;
+
+  protected GenericSearchEmailService(MailboxRepository mailboxRepository) {
+    this.mailboxRepository = mailboxRepository;
+  }
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/GenericSearchEmailService.java:20]:
```suggestion
  private final ProvideEmailUseCase provideEmailUseCase;

  protected GenericSearchEmailService(ProvideEmailUseCase provideEmailUseCase) {
    this.provideEmailUseCase= provideEmailUseCase;
  }
```
Theoretically you could also use this service here. In this way you dont have to iterate through each mailbox but you can use the provideEmailUseCase.provideEmails() method

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/App.java
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/App.java:1]: Perfect!

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/subject/SearchSubjectEmailService.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.Mailbox;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.MailboxRepository;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.enums.MailboxType;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.exceptions.MailboxLoadingException;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.exceptions.MailboxSavingException;
+import de.dhbw.karlsruhe.students.mailflow.core.domain.email.value_objects.Address;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SearchSubjectEmailService implements SearchSubjectEmailUseCase {
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/subject/SearchSubjectEmailService.java:13]:
```suggestion
public class SearchSubjectEmailService implements SearchEmailUseCase {
```

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/subject/SearchSubjectEmailService.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        .filter(email -> email.getSubject().subject().contains(subject))
+        .toList();
+  }
+
+  @Override
+  public List<Email> filterEmailsBySubject(List<Email> emails, String subject) {
+    return emails.stream()
+        .filter(email -> email.getSubject().subject().contains(subject))
+        .toList();
+  }
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/subject/SearchSubjectEmailService.java:40]:
```suggestion

```
You dont need this

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/subject/SearchSubjectEmailUseCase.java
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/application/email/searchemail/subject/SearchSubjectEmailUseCase.java:1]: This interface can be removed

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/cli/BaseCLIPrompt.java
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/cli/BaseCLIPrompt.java:1]: Looks good!

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/cli/usecases/searchemails/SearchEmailTypesCLIPrompt.java
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/cli/usecases/searchemails/SearchEmailTypesCLIPrompt.java:1]: Looks good!

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/cli/usecases/searchemails/SearchEmailTypesCLIPrompt.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    BaseCLIPrompt action = showActionMenuPrompt();
+    action.start();
+  }
+
+  private BaseCLIPrompt showActionMenuPrompt() {
+    Map<String, BaseCLIPrompt> promptMap = new LinkedHashMap<>();
+    promptMap.put(
+        "Subject",
+        new SearchEmailCLIPrompt(
+            this, authUseCase, searchEmails.searchSubjectEmailUseCase()));
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/cli/usecases/searchemails/SearchEmailTypesCLIPrompt.java:33]: You probably also need to add the provideEmailsUseCase

**Review by Jonas-Karl — CHANGES_REQUESTED** [review]

**Review by Jonas-Karl — APPROVED** [review]
Wonderfull

**Review by jens1o — APPROVED** [review]

**Pull request merged**